edalize/nextpnr.py

```diff
@@ -0,0 +1,124 @@
+                        "desc": "Output file format. Legal values are *fasm*",
+                    },
+                    {
+                        "name": "nextpnr_as_subtool",
+                        "type": "bool",
+                        "desc": "Determines if nextpnr is run as a part of bigger toolchain, or as a standalone tool",
+                    },
+                ],
+                "lists": [
+                    {
+                        "name": "nextpnr_impl_options",
+                        "type": "String",
+                        "desc": "Additional options for the implementation command",
+                    },
+                ],
+            }
+
+            combined_members = nextpnr_help["members"]
+            combined_lists = nextpnr_help.get("lists", [])
+            yosys_members = yosys_help["members"]
+            yosys_lists = yosys_help["lists"]
+
+            combined_members.extend(
+                m
+                for m in yosys_members
+                if m["name"] not in [i["name"] for i in combined_members]
+            )
+            combined_lists.extend(
+                l
+                for l in yosys_lists
+                if l["name"] not in [i["name"] for i in combined_lists]
+            )
+
+            return {
+                "description": "Open source Place and Route tool targeting many different FPGAs",
+                "members": combined_members,
+                "lists": combined_lists,
+            }
+
+    def configure_main(self):
+        (src_files, incdirs) = self._get_fileset_files()
+
+        yosys_synth_options = self.tool_options.get('yosys_synth_options', '')
+        yosys_edam = {
+                'files'         : self.files,
+                'name'          : self.name,
+                'toplevel'      : self.toplevel,
+                'parameters'    : self.parameters,
+                'tool_options'  : {'yosys' : {
+                                        'arch' : 'xilinx',
+                                        'yosys_synth_options' : yosys_synth_options,
+                                        'yosys_as_subtool' : True,
+                                        }
+                                }
+                }
+
+        yosys = getattr(import_module("edalize.yosys"), "Yosys")(
+            yosys_edam, self.work_root
+        )
+        yosys.configure()
+
+        part_of_toolchain = self.tool_options.get("nextpnr_as_subtool", False)
+
+        file_table = []
+
+        output_format = self.tool_options.get("output_format", "fasm")
+        arch = self.tool_options.get("arch", "xilinx")
+
+        synth_design = None
+        chipdb = None
+        xdc = None
+
+        for f in src_files:
+            if f.file_type in ["synthJson"]:
+                synth_design = f.name
+            elif f.file_type in ["bba"]:
+                chipdb = f.name
+            elif f.file_type in ["xdc"]:
+                xdc = f.name
```

Comment:
Does the backend only support one xdc file? If multiple are used, this will silently overwrite the first ones and only use the last. Either warn/error if there are more than one, or merge all files into a single file

Reply:
From `nextpnr-xilinx` help:
```  --xdc arg                 XDC-style constraints file ```
I assume it is just one file. I added the [get_file_type_name](https://github.com/olofk/edalize/pull/193/files#diff-f2231ab9d2b0eca263ed813cc4ba1ee5R62) function to log warnings whenever there are multiple files provided.

Reply:
Cool. At some point we can pull this out to a common file since similar checks are done in other backends. But this is fine for now